src/linter/block.go

```diff
@@ -1361,6 +1362,27 @@ func (b *BlockWalker) handleArrayItems(arr node.Node, items []*expr.ArrayItem) b
 			constKey = true
 		case *scalar.Lnumber:
 			key = k.Value
+			constKey = true
+		case *expr.ConstFetch:
+			_, info, ok := solver.GetConstant(b.r.ctx.st, k.Constant)
+
+			if !ok {
+				continue
+			}
+
+			key = info.Value.Value
+
+			if info.Value.Type == meta.Float {
+				val, err := strconv.ParseFloat(key, 64)
+				if err != nil {
+					continue
+				}
+				value := int64(math.Floor(val))
+				key = fmt.Sprint(value)
```

Comment:
You might want to test this code path.

From the PHP manual:
> Floats are also cast to integers, which means that the fractional part will be truncated. E.g. the key 8.7 will actually be stored under 8. 

This gives a hint that values are "truncated", not floor'ed.

For some inputs, it behaves the same, but:
```
int(floor(-1.51)) => int(-2.0) => -2
int(-1.51) => -1
```
PHP will do the latter:
```php
var_dump([-1.5 => 1]);
```
```
array(1) {
  [-1]=>
  int(1)
}
```
I would suggest skipping the floor step.
It's also a good idea to add tests that check this corner case behavior for `-1.51`.